File: fortran/src/main/java/de/irt/jfor/R4.java
package de.irt.jfor;

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  19.10.2017 13:53
 * modified by: $Author$
 * modified on: $Date$
 */
public class R4 {

    public float v = 0;

    public R4(float value) {
        v = value;
    }

    public String toString() {
        return Float.toString(v);
    }

    public static R4 of() {
        return new R4(0);
    }

    public static R4 of(float val) {
        return new R4(val);
    }

    public static R4 of(double val) {
        return new R4((float)val);
    }

}

File: fortran/src/main/java/de/irt/kfor/R4.kt
package de.irt.kfor

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  09.01.2019 09:27
 * modified by: $Author$
 * modified on: $Date$
 */
interface R4 {

    companion object {
        operator fun invoke(v : Double = 0.0) = Ref(v)
        fun arr(ni : Int) = Arr(ni)
        fun arr() = arr(0)
        fun mat(ni : Int,nj : Int) = Mat(ni, nj)
        fun cub(ni : Int,nj : Int, nk : Int) = Cub(ni, nj, nk)
    }

    data class Arr(override val ni : Int) : RArr {
        val arr = FloatArray(len)
        override fun get(i : Int) = arr[index(i)].toDouble()
        override fun set(i : Int, v : Double) {
            arr[index(i)] = v.toFloat()
        }

        override fun allocate(ni : Int) = if(this.len==ni) this else Arr(ni)
    }

    data class Mat (override val ni : Int, override val nj : Int) : RMat {
        val arr = FloatArray(len)
        override fun get(i : Int, j : Int) = arr[index(i, j)].toDouble()
        override fun set(i : Int, j : Int, v : Double) {
            arr[index(i, j)] = v.toFloat()
        }
    }

    data class Cub (override val ni : Int, override val nj : Int, override val nk : Int) : RCub {
        val arr = FloatArray(len)
        override fun get(i : Int, j : Int, k : Int) = arr[index(i,j,k)].toDouble()
        override fun set(i : Int, j : Int, k : Int, v : Double) {
            arr[index(i,j,k)] = v.toFloat()
        }
    }
}
